// appium-remote-debugger/build/lib/mixins/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.events = void 0;
exports.addClientEventListener = addClientEventListener;
exports.removeClientEventListener = removeClientEventListener;
// event emitted publically
exports.events = {
    EVENT_PAGE_CHANGE: 'remote_debugger_page_change',
    EVENT_FRAMES_DETACHED: 'remote_debugger_frames_detached',
    EVENT_DISCONNECT: 'remote_debugger_disconnect',
};
/**
 * Keep track of the client event listeners so they can be removed
 *
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {string} eventName
 * @param {(event: import('@appium/types').StringRecord) => any} listener
 * @returns {void}
 */
function addClientEventListener(eventName, listener) {
    var _a;
    (_a = this._clientEventListeners)[eventName] ?? (_a[eventName] = []);
    this._clientEventListeners[eventName].push(listener);
    this.requireRpcClient().on(eventName, listener);
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {string} eventName
 * @returns {void}
 */
function removeClientEventListener(eventName) {
    for (const listener of (this._clientEventListeners[eventName] || [])) {
        this.requireRpcClient().off(eventName, listener);
    }
}
exports.default = exports.events;
