// appium-remote-debugger/build/lib/mixins/message-handlers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.onPageChange = onPageChange;
exports.onAppConnect = onAppConnect;
exports.onAppDisconnect = onAppDisconnect;
exports.onAppUpdate = onAppUpdate;
exports.onConnectedDriverList = onConnectedDriverList;
exports.onCurrentState = onCurrentState;
exports.onConnectedApplicationList = onConnectedApplicationList;
const events_1 = __importDefault(require("./events"));
const utils_1 = require("../utils");
const lodash_1 = __importDefault(require("lodash"));
/*
 * Generic callbacks used throughout the lifecycle of the Remote Debugger.
 * These will be added to the prototype.
 */
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {Error?} err
 * @param {string} appIdKey
 * @param {Record<string, any>} pageDict
 * @returns {Promise<void>}
 */
async function onPageChange(err, appIdKey, pageDict) {
    if (lodash_1.default.isEmpty(pageDict)) {
        return;
    }
    const pageArray = (0, utils_1.pageArrayFromDict)(pageDict);
    await useAppDictLock.bind(this)((/** @type {() => void} */ done) => {
        try {
            // save the page dict for this app
            if (this.appDict[appIdKey]) {
                if (this.appDict[appIdKey].pageArray) {
                    if (this.appDict[appIdKey].pageArray.resolve) {
                        // pageDict is a pending promise, so resolve
                        this.appDict[appIdKey].pageArray.resolve();
                    }
                    else {
                        // we have a pre-existing pageDict
                        if (lodash_1.default.isEqual(this.appDict[appIdKey].pageArray, pageArray)) {
                            this.log.debug(`Received page change notice for app '${appIdKey}' ` +
                                `but the listing has not changed. Ignoring.`);
                            return done();
                        }
                    }
                }
                // keep track of the page dictionary
                this.appDict[appIdKey].pageArray = pageArray;
            }
        }
        finally {
            done();
        }
    });
    if (this._navigatingToPage) {
        // in the middle of navigating, so reporting a page change will cause problems
        return;
    }
    this.log.debug(`Page changed: ${(0, utils_1.simpleStringify)(pageDict, true)}`);
    this.emit(events_1.default.EVENT_PAGE_CHANGE, {
        appIdKey: appIdKey.replace('PID:', ''),
        pageArray,
    });
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {Error?} err
 * @param {Record<string, any>} dict
 * @returns {Promise<void>}
 */
async function onAppConnect(err, dict) {
    const appIdKey = dict.WIRApplicationIdentifierKey;
    this.log.debug(`Notified that new application '${appIdKey}' has connected`);
    await useAppDictLock.bind(this)((/** @type {() => void} */ done) => {
        try {
            updateAppsWithDict.bind(this)(dict);
        }
        finally {
            done();
        }
    });
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {Error?} err
 * @param {Record<string, any>} dict
 * @returns {void}
 */
function onAppDisconnect(err, dict) {
    const appIdKey = dict.WIRApplicationIdentifierKey;
    this.log.debug(`Application '${appIdKey}' disconnected. Removing from app dictionary.`);
    this.log.debug(`Current app is '${this.appIdKey}'`);
    // get rid of the entry in our app dictionary,
    // since it is no longer available
    delete this.appDict[appIdKey];
    // if the disconnected app is the one we are connected to, try to find another
    if (this.appIdKey === appIdKey) {
        this.log.debug(`No longer have app id. Attempting to find new one.`);
        this.appIdKey = (0, utils_1.getDebuggerAppKey)(/** @type {string} */ (this.bundleId), this.appDict);
    }
    if (!this.appDict) {
        // this means we no longer have any apps. what the what?
        this.log.debug('Main app disconnected. Disconnecting altogether.');
        this.connected = false;
        this.emit(events_1.default.EVENT_DISCONNECT, true);
    }
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {Error?} err
 * @param {Record<string, any>} dict
 * @returns {Promise<void>}
 */
async function onAppUpdate(err, dict) {
    await useAppDictLock.bind(this)((/** @type {() => void} */ done) => {
        try {
            updateAppsWithDict.bind(this)(dict);
        }
        finally {
            done();
        }
    });
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {Error?} err
 * @param {Record<string, any>} drivers
 * @returns {void}
 */
function onConnectedDriverList(err, drivers) {
    this.connectedDrivers = drivers.WIRDriverDictionaryKey;
    this.log.debug(`Received connected driver list: ${JSON.stringify(this.connectedDrivers)}`);
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {Error?} err
 * @param {Record<string, any>} state
 * @returns {void}
 */
function onCurrentState(err, state) {
    this.currentState = state.WIRAutomationAvailabilityKey;
    // This state changes when 'Remote Automation' in 'Settings app' > 'Safari' > 'Advanced' > 'Remote Automation' changes
    // WIRAutomationAvailabilityAvailable or WIRAutomationAvailabilityNotAvailable
    this.log.debug(`Received connected automation availability state: ${JSON.stringify(this.currentState)}`);
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {Error?} err
 * @param {Record<string, any>} apps
 * @returns {Promise<void>}
 */
async function onConnectedApplicationList(err, apps) {
    this.log.debug(`Received connected applications list: ${lodash_1.default.keys(apps).join(', ')}`);
    // translate the received information into an easier-to-manage
    // hash with app id as key, and app info as value
    let newDict = {};
    for (const dict of lodash_1.default.values(apps)) {
        const [id, entry] = (0, utils_1.appInfoFromDict)(dict);
        if (this.skippedApps.includes(entry.name)) {
            continue;
        }
        newDict[id] = entry;
    }
    // update the object's list of apps
    await useAppDictLock.bind(this)((/** @type {() => void} */ done) => {
        try {
            lodash_1.default.defaults(this.appDict, newDict);
        }
        finally {
            done();
        }
    });
}
/**
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {(done: () => any) => any} fn
 * @returns {Promise<any>}
 */
async function useAppDictLock(fn) {
    return await this._lock.acquire('appDict', fn);
}
/**
 *
 * @this {import('../remote-debugger').RemoteDebugger}
 * @param {import('@appium/types').StringRecord} dict
 * @returns {void}
 */
function updateAppsWithDict(dict) {
    // get the dictionary entry into a nice form, and add it to the
    // application dictionary
    this.appDict = this.appDict || {};
    let [id, entry] = (0, utils_1.appInfoFromDict)(dict);
    if (this.appDict[id]) {
        // preserve the page dictionary for this entry
        entry.pageArray = this.appDict[id].pageArray;
    }
    this.appDict[id] = entry;
    // add a promise to get the page dictionary
    if (lodash_1.default.isUndefined(entry.pageArray)) {
        entry.pageArray = (0, utils_1.deferredPromise)();
    }
    // try to get the app id from our connected apps
    if (!this.appIdKey) {
        this.appIdKey = (0, utils_1.getDebuggerAppKey)(/** @type {string} */ (this.bundleId), this.appDict);
    }
}
